barretenberg/cpp/src/barretenberg/boomerang_value_detection/graph.cpp

```diff
@@ -0,0 +1,824 @@
+                bool find_out = find_position(real_out_idx);
+                if (find_out) {
+                    variables_in_one_gate.erase(real_out_idx);
+                }
+            }
+        }
+    }
+}
+
+/**
+ * @brief this method removes false cases plookup variables from variables in one gate
+ * @tparam FF
+ * @param ultra_circuit_builder
+ * @param variables_in_one_gate
+ */
+
+template <typename FF>
+inline void Graph_<FF>::remove_unnecessary_plookup_variables(bb::UltraCircuitBuilder& ultra_circuit_builder,
+                                                             std::unordered_set<uint32_t>& variables_in_one_gate)
+{
+    auto& lookup_block = ultra_circuit_builder.blocks.lookup;
+    if (lookup_block.size() > 0) {
+        for (size_t i = 0; i < lookup_block.size(); i++) {
+            process_current_plookup_gate(ultra_circuit_builder, variables_in_one_gate, i);
+        }
+    }
+}
+
+/**
+ * @brief this method returns a final set of variables that were in one gate
+ * @tparam FF
+ * @param ultra_circuit_builder
+ * @return std::unordered_set<uint32_t>
+ */
+
+template <typename FF>
+std::unordered_set<uint32_t> Graph_<FF>::show_variables_in_one_gate(bb::UltraCircuitBuilder& ultra_circuit_builder)
+{
+    std::unordered_set<uint32_t> variables_in_one_gate;
+    for (const auto& pair : variables_gate_counts) {
+        bool is_not_constant_variable = check_is_not_constant_variable(ultra_circuit_builder, pair.first);
+        if (pair.second == 1 && pair.first != 0 && is_not_constant_variable) {
+            variables_in_one_gate.insert(pair.first);
+        }
+    }
+    auto range_lists = ultra_circuit_builder.range_lists;
+    std::unordered_set<uint32_t> decompose_varialbes;
+    for (auto& pair : range_lists) {
+        for (auto& elem : pair.second.variable_indices) {
+            bool is_not_constant_variable = check_is_not_constant_variable(ultra_circuit_builder, elem);
+            if (variables_gate_counts[ultra_circuit_builder.real_variable_index[elem]] == 1 &&
+                is_not_constant_variable) {
+                decompose_varialbes.insert(ultra_circuit_builder.real_variable_index[elem]);
+            }
+        }
+    }
+    remove_unnecessary_decompose_variables(ultra_circuit_builder, variables_in_one_gate, decompose_varialbes);
+    remove_unnecessary_plookup_variables(ultra_circuit_builder, variables_in_one_gate);
+    return variables_in_one_gate;
+}
+
+/**
+ * @brief this method returns connected component with a given index and size of this component
+ *
+ * @param connected_components
+ * @param index
+ * @return std::pair<std::vector<uint32_t>, size_t>
+ */
+
+std::pair<std::vector<uint32_t>, size_t> get_connected_component_with_index(
+    const std::vector<std::vector<uint32_t>>& connected_components, size_t index)
+{
+    auto connected_component = connected_components[index];
+    auto size = connected_component.size();
+    return std::make_pair(connected_component, size);
```

Comment:
Why do you need to return a pair?

Reply:
Sometimes information about a size of connected component can be useful for debugging. For example, if connected component consist from 1, 2, 3, 4 variables, it means that there was some error in a creation of the graph. And sometimes I want to print this connected component and find a gate for these variables

Reply:
ok